pkg/ccl/changefeedccl/changefeed_test.go

```diff
@@ -5831,39 +5831,35 @@ func TestChangefeedEndTime(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	testFn := func(t *testing.T, db *gosql.DB, f cdctest.TestFeedFactory) {
+		knobs := f.Server().TestingKnobs().
+			DistSQL.(*execinfra.TestingKnobs).
+			Changefeed.(*TestingKnobs)
+		endTimeReached := make(chan struct{})
+		knobs.FeedKnobs.EndTimeReached = func() bool {
+			select {
+			case <-endTimeReached:
+				return true
+			default:
+				return false
+			}
+		}
+
 		sqlDB := sqlutils.MakeSQLRunner(db)
 
 		sqlDB.Exec(t, "CREATE TABLE foo (a INT PRIMARY KEY)")
 		sqlDB.Exec(t, "INSERT INTO foo VALUES (1), (2), (3)")
 
-		var tsAfterInitialInsert string
-		sqlDB.QueryRow(t, "SELECT (cluster_logical_timestamp() + 10000000000)").Scan(&tsAfterInitialInsert)
-		feed := feed(t, f, "CREATE CHANGEFEED FOR foo WITH end_time = $1", tsAfterInitialInsert)
-
-		time.Sleep(10 * time.Second)
-		sqlDB.Exec(t, "INSERT INTO foo VALUES (4), (5), (6)")
+		fakeEndTime := f.Server().Clock().Now().Add(int64(time.Hour), 0).AsOfSystemTime()
+		feed := feed(t, f, "CREATE CHANGEFEED FOR foo WITH end_time = $1", fakeEndTime)
+		defer closeFeed(t, feed)
 
-		seenMoreMessages := false
-		g := ctxgroup.WithContext(context.Background())
-		g.Go(func() error {
-			assertPayloads(t, feed, []string{
-				`foo: [1]->{"after": {"a": 1}}`,
-				`foo: [2]->{"after": {"a": 2}}`,
-				`foo: [3]->{"after": {"a": 3}}`,
-			})
-			for {
-				_, err := feed.Next()
-				if err != nil {
-					return err
-				}
-				seenMoreMessages = true
-			}
+		assertPayloads(t, feed, []string{
+			`foo: [1]->{"after": {"a": 1}}`,
+			`foo: [2]->{"after": {"a": 2}}`,
+			`foo: [3]->{"after": {"a": 3}}`,
 		})
-		defer func() {
-			closeFeed(t, feed)
-			_ = g.Wait()
-			require.False(t, seenMoreMessages)
-		}()
+
+		close(endTimeReached)
```

Comment:
I feel like we're no longer testing enough of the endTime implementation, since now we're skipping over all the code around setting the scanboundary at the right timestamp and reacting appropriately, we're only verifying that the changefeed ends with Success if end_time was set since the knob does all of the initial work of stopping the feed.  We could miss bugs where we accidentally emit extra messages that were emitted after the endTime.

It'd also be nice to sneak in a schema change after creating the changefeed but before the endTime just to introduce another scan boundary that shouldn't be invalidated by the current scanboundary-based endtime code.

Reply:
Well, perhaps it's not as thorough; but I'm not concerned.  First, sleep, particularly sleep for 10 seconds is really not great.  Secondly, we still do set the boundary -- it's just triggering scan boundary at different time than it would otherwise. 

>  We could miss bugs where we accidentally emit extra messages that were emitted after the endTime.

Perhaps; but also consider the fact that the existing check for those messages is racy -- it happens to work because
of how our fake test feed implementations work.  So, I don't really feel it's loosing that much coverage.

> It'd also be nice to sneak in a schema change after creating the changefeed but before the endTime just to introduce another scan boundary that shouldn't be invalidated by the current scanboundary-based endtime code.

Sure, we can improve this test... I'm just trying to fix what's broken right now - namely the 10 second sleep.

Reply:
I do still feel uncomfortable about lowering the test coverage even though it works via our testfeed details + sleep, but not a strong opinion that'd block merge